### PR TITLE
RFC: More methods for Irrational

### PR DESCRIPTION
**Pull request by perrutquist**
The purpose of this PR is to make `Irrational` work with many more functions, including those mentioned in #31949 #26701 and #22878.

The main way that this is accomplished is by defining
```
zero(::AbstractIrrational) = false
zero(::Type{<:AbstractIrrational}) = false
one(::Type{<:AbstractIrrational}) = true
```

The documentation says that these functions should return the additive/multiplicative identities *for* the type `T` - not *of* the type T.

In order to make `false` and `true` into additive/multiplicative identities, this PR also re-defines `Irrational`-`Bool` addition, subtraction and multiplication.
This is technically breaking if someone is currently using `true * pi` or `pi + false` as a way of forcing conversion to `Float64`. (The alternatives `1 * pi` and `0 + pi` continue to work as before.)
The new +, - and * methods are not type-stable, which the old ones were. On the other hand, Julia is now much better at type inference than it was when the original methods were written.

With this, a whole range of functions start to work, as the only reason they were throwing before was a call to `one` or `zero`. To give one example, `sec(x::Number)` is defined as `one(x)/cos(x)` which now works.

This PR also extends the list of two-argument functions that convert to Float64 when given two identical `Irrational` arguments. The intent is to make all "math-like" functions in `Base` work with `Irrational`. (This should mitigate the effect of `true*pi` now returning `pi` rather than a `Float64`.)

Finally, some minor fixes: `widemul` of two `Irrational`s now returns a `BigFloat` same as `widemul` of `Float64`s, and  `sin(pi)` and `tan(pi)` now return exactly `0.0`.

### CODE REVIEW

**Comment by perrutquist** [comment]
Another effect that should be mentioned is that `complex(pi)` (previously) an error, will now return a `Complex{Float64}` because of how `(pi, zero(pi))` is promoted.

**Comment by perrutquist** [comment]
...and for that reason it seemed to make sense to also make `Complex(pi, pi)` a `Complex{Float64}` rather than `Complex{Irrational{:π}}`. It would be a very odd special case.

**Comment by ararslan** [comment]
> The documentation says that these functions should return the additive/multiplicative identities for the type T - not of the type T.

I think that's taking the documentation a bit overly literal in this case. I don't know of any other cases where `zero`/`one` does not return value of the input type. Indeed, it's assumed in a number of places, especially when determining the element type for a container.

**Comment by stevengj** [comment]
> I don't know of any other cases where `zero`/`one` does not return value of the input type.

`oneunit(T)` usually returns something of type `T`, but `one(T)` does not for dimensionful types.

**Comment by mohamed82008** [comment]
I am running into issues with `DiffRules` and `Distributions` that this PR can fix. Are people here for or against the proposed changes? If only a rebase is needed, I can go ahead and do that.

**Comment by mohamed82008** [comment]
I only care about `one` and `zero` for irrationals.

File: base/irrationals.jl [hunk 1]
```diff
@@ -132,14 +139,43 @@ hash(x::Irrational, h::UInt) = 3*objectid(x) - h
...
++(x::Bool, y::AbstractIrrational) = x ? 1.0 + y : y
++(x::AbstractIrrational, y::Bool) = y + x
+-(x::AbstractIrrational, y::Bool) = y ? x - 1.0 : x
 
 round(x::Irrational, r::RoundingMode) = round(float(x), r)
 
+Rational(x::AbstractIrrational) = rationalize(x)
+rationalize(x::Irrational{T}) where T = throw(ArgumentError("Type must be specified. Use rationalize(Int, $T) to obtain a Rational{Int} approximation to the irrational constant $T."))
+Math.mod2pi(x::AbstractIrrational) = Float64(mod2pi(big(x)))
+Math.rem2pi(x::AbstractIrrational, m) = Float64(rem2pi(big(x), m))
```
Comment (stevengj) [inline base/irrationals.jl]: Shouldn't use `BigFloat` arithmetic at runtime.

**Comment by stevengj** [comment]
I'm okay with `one` and `zero`.  The other functions are mostly fine but might require some more thought.  My suggestion would be to make a separate PR with just `one` and `zero` if that's all you need for now, since that can be merged quickly.

**Comment by mohamed82008** [comment]
Sounds good, will do.

**Comment by mohamed82008** [comment]
Done.

File: base/irrationals.jl (hunk 1 above)
Comment (perrutquist) [reply base/irrationals.jl]: Good point. I will turn these into generated functions.

**Comment by perrutquist** [comment]
I'm very sorry. I tried to rebase this on the latest master, and I obviously messed up.  Now it looks like I've modified 1658 files. Git and I are not always friends. Will try to fix this.

File: base/irrationals.jl [hunk 2]
```diff
@@ -151,14 +158,38 @@ zero(::Type{<:AbstractIrrational}) = false
...
++(x::Bool, y::AbstractIrrational) = x ? 1.0 + y : y
++(x::AbstractIrrational, y::Bool) = y + x
+-(x::AbstractIrrational, y::Bool) = y ? x - 1.0 : x
 
 round(x::Irrational, r::RoundingMode) = round(float(x), r)
 
+Rational(x::AbstractIrrational) = rationalize(x)
+rationalize(x::Irrational{T}) where T = throw(ArgumentError("Type must be specified. Use rationalize(Int, $T) to obtain a Rational{Int} approximation to the irrational constant $T."))
+@generated Math.mod2pi(x::AbstractIrrational) = Float64(mod2pi(big(x())))
+@generated Math.rem2pi(x::AbstractIrrational, r::RoundingMode) = Float64(rem2pi(big(x()), r()))
```
Comment (vtjnash) [inline base/irrationals.jl:179]:
```suggestion
Math.mod2pi(x::AbstractIrrational) = Float64(mod2pi(big(x)))
Math.rem2pi(x::AbstractIrrational, r::RoundingMode) = Float64(rem2pi(big(x), r))
```

File: base/irrationals.jl [hunk 3]
```diff
@@ -151,14 +158,38 @@ zero(::Type{<:AbstractIrrational}) = false
...
+oneunit(T::Type{<:AbstractIrrational}) = throw(ArgumentError("The number one cannot be of type $T"))
+oneunit(x::T) where {T <: AbstractIrrational} = oneunit(T)
+
 -(x::AbstractIrrational) = -Float64(x)
-for op in Symbol[:+, :-, :*, :/, :^]
+for op in Symbol[:+, :-, :*, :/, :^, :cld, :div, :divrem, :fld, :mod, :mod1, :rem, :UnitRange]
     @eval $op(x::AbstractIrrational, y::AbstractIrrational) = $op(Float64(x),Float64(y))
 end
-*(x::Bool, y::AbstractIrrational) = ifelse(x, Float64(y), 0.0)
+*(x::Bool, y::AbstractIrrational) = x ? y : zero(y)
```
Comment (JeffBezanson) [inline base/irrationals.jl:168]: Type unstable --- I'm not sure this is worth it.

File: base/irrationals.jl [hunk 4]
```diff
@@ -45,6 +51,7 @@ promote_rule(::Type{S}, ::Type{T}) where {S<:AbstractIrrational,T<:Number} = pro
 AbstractFloat(x::AbstractIrrational) = Float64(x)::Float64
 Float16(x::AbstractIrrational) = Float16(Float32(x)::Float32)
 Complex{T}(x::AbstractIrrational) where {T<:Real} = Complex{T}(T(x))
+Complex(x::AbstractIrrational, y::AbstractIrrational) = Complex(Float64(x), Float64(y))
```
Comment (vtjnash) [inline base/irrationals.jl:54]: Seems unnecessary?
```
julia> Complex(π, ℯ)
3.141592653589793 + 2.718281828459045im
```
```suggestion
```

File: base/irrationals.jl [hunk 5]
```diff
@@ -151,14 +158,38 @@ zero(::Type{<:AbstractIrrational}) = false
...
+# but sin(pi) should return zero. (The same applies to tan(pi).)
+sin(::Irrational{:π}) = 0.0
+tan(::Irrational{:π}) = 0.0
+Math.sincos(::Irrational{:π}) = (0.0, -1.0)
+
+sign(x::AbstractIrrational) = sign(Float64(x))
+
+widemul(x::AbstractIrrational, y::AbstractIrrational) = big(x)*big(y)
+
+(::Colon)(x::AbstractIrrational, y::AbstractIrrational) = Float64(x):Float64(y)
```
Comment (vtjnash) [inline base/irrationals.jl:191]: Seems to already work without this?
```suggestion
```
```
julia> ℯ:π
2.718281828459045:1.0:2.718281828459045
```

**Comment by vtjnash** [comment]
We discussed this on triage, and felt that it is better to be explicit about the transforms for irrational numbers into the appropriate computational types. Thanks for the PR however. I know it is sad to see one not get merged, but it happens to all of us.

**Pull request closed without merging**